File: packages/runtime/src/agent-telemetry.ts
import type {
  AgentAttemptTelemetry,
  AgentAttemptTelemetryState,
  AgentContextUsage,
  AgentIoPreview,
  AgentTelemetry,
  AgentToolCallTelemetry
} from "./types.js";

const DEFAULT_PREVIEW_EDGE_BYTES = 8 * 1024;
const DEFAULT_MAX_TOOL_CALLS = 200;
const DEFAULT_FLUSH_INTERVAL_MS = 1_000;
const FINAL_TOOL_STATUSES = new Set(["completed", "failed", "cancelled"]);

export interface AgentTelemetryAccumulatorOptions {
  attempt: number;
  inputText: string;
  inputArtifactRef?: string;
  startedAt?: string;
  now?: () => number;
  previewEdgeBytes?: number;
  maxToolCalls?: number;
  flushIntervalMs?: number;
  onTelemetry?: (attempt: AgentAttemptTelemetry) => void;
}

export class AgentTelemetryAccumulator {
  private readonly attempt: number;
  private readonly now: () => number;
  private readonly previewEdgeBytes: number;
  private readonly maxToolCalls: number;
  private readonly flushIntervalMs: number;
  private readonly onTelemetry?: (attempt: AgentAttemptTelemetry) => void;
  private readonly startedAt: string;
  private readonly seenToolIds = new Set<string>();
  private readonly tools = new Map<string, MutableToolCall>();
  private partialLine = "";
  private response = "";
  private stopReason: string | undefined;
  private latestContext: AgentContextUsage | undefined;
  private outputArtifactRef: string | undefined;
  private lastPublishedAt = 0;
  private toolSeq = 0;
  private totalToolCallCount = 0;

  constructor(options: AgentTelemetryAccumulatorOptions) {
    this.attempt = options.attempt;
    this.now = options.now ?? Date.now;
    this.previewEdgeBytes = options.previewEdgeBytes ?? DEFAULT_PREVIEW_EDGE_BYTES;
    this.maxToolCalls = options.maxToolCalls ?? DEFAULT_MAX_TOOL_CALLS;
    this.flushIntervalMs = options.flushIntervalMs ?? DEFAULT_FLUSH_INTERVAL_MS;
    this.onTelemetry = options.onTelemetry;
    this.startedAt = options.startedAt ?? new Date(this.now()).toISOString();
    this.input = buildPreview(options.inputText, this.previewEdgeBytes, options.inputArtifactRef);
  }

  private readonly input: AgentIoPreview;

  append(chunk: string): void {
    if (chunk.length === 0) return;
    const text = this.partialLine + chunk;
    const lines = text.split("\n");
    this.partialLine = lines.pop() ?? "";
    for (const line of lines) this.parseLine(line);
  }

  flush(): void {
    if (this.partialLine.trim().length > 0) this.parseLine(this.partialLine);
    this.partialLine = "";
    this.publish("running", true);
  }

  responseText(): string {
    return this.response;
  }

  setResponseText(text: string): void {
    this.response = text;
  }

  finalStopReason(): string | undefined {
    return this.stopReason;
  }

  context(): AgentContextUsage | undefined {
    return this.latestContext;
  }

  setOutputArtifactRef(ref: string | undefined): void {
    this.outputArtifactRef = ref;
  }

  snapshot(state: AgentAttemptTelemetryState = "running", completedAt?: string): AgentAttemptTelemetry {
    const updatedAt = completedAt ?? new Date(this.now()).toISOString();
    const recentCalls = [...this.tools.values()]
      .sort((a, b) => b.lastSeq - a.lastSeq)
      .map(({ lastSeq: _lastSeq, ...tool }) => tool);
    return {
      attempt: this.attempt,
      state,
      startedAt: this.startedAt,
      updatedAt,
      completedAt,
      context: this.latestContext,
      input: this.input,
      output: this.response.length > 0
        ? buildPreview(this.response, this.previewEdgeBytes, this.outputArtifactRef)
        : this.outputArtifactRef
          ? buildPreview("", this.previewEdgeBytes, this.outputArtifactRef)
          : undefined,
      tools: {
        totalToolCallCount: this.totalToolCallCount,
        droppedToolCallCount: Math.max(0, this.totalToolCallCount - recentCalls.length),
        recentCalls
      }
    };
  }

  private parseLine(line: string): void {
    const trimmed = line.trim();
    if (!trimmed) return;
    const parsed = parseObject(trimmed);
    if (!parsed) return;

    this.captureResult(parsed);
    const update = extractUpdate(parsed);
    if (!update) return;

    const kind = update.sessionUpdate;
    if (kind === "agent_message_chunk") {
      const text = readContentText(update.content);
      if (text) {
        this.response += text;
        this.publish("running", false);
      }
      return;
    }

    if (kind === "usage_update") {
      const used = readNonNegativeNumber(update.used);
      const size = readNonNegativeNumber(update.size);
      if (used !== undefined && size !== undefined) {
        // acpx sends used=0 at the start of every LLM API call before the
        // response arrives with real token counts.  Overwriting a known
        // non-zero measurement with 0 would produce misleading telemetry
        // (e.g. "context=0/200k") when a failed attempt's last update was
        // an initial allocation.  Preserve the previous `used` if the new
        // one is 0 and we already have a real measurement.
        if (used === 0 && this.latestContext && this.latestContext.used > 0) {
          this.latestContext = { used: this.latestContext.used, size, updatedAt: new Date(this.now()).toISOString() };
        } else {
          this.latestContext = { used, size, updatedAt: new Date(this.now()).toISOString() };
        }
        this.publish("running", false);
      }
      return;
    }

    if (isToolUpdate(update)) {
      const changed = this.captureTool(update);
      this.publish("running", changed);
    }
  }

  private captureResult(obj: Record<string, unknown>): void {
    const result = readRecord(obj.result);
    if (typeof result?.stopReason === "string") this.stopReason = result.stopReason;
  }

  private captureTool(update: Record<string, unknown>): boolean {
    const toolCallId = readString(update.toolCallId);
    if (!toolCallId) return false;

    const now = new Date(this.now()).toISOString();
    const wasSeen = this.seenToolIds.has(toolCallId);
    if (!wasSeen) {
      this.seenToolIds.add(toolCallId);
      this.totalToolCallCount += 1;
    }

    const previous = this.tools.get(toolCallId);
    const status = readString(update.status);
    const current: MutableToolCall = previous ?? {
      toolCallId,
      startedAt: now,
      updatedAt: now,
      lastSeq: 0
    };
    const statusChanged = status !== undefined && status !== previous?.status;

    current.updatedAt = now;
    current.lastSeq = ++this.toolSeq;
    const title = readString(update.title);
    const kind = readString(update.kind);
    const toolName = readString(readRecord(readRecord(update._meta)?.claudeCode)?.toolName);
    if (title) current.title = title;
    if (kind) current.kind = kind;
    if (toolName) current.toolName = toolName;
    if (status) {
      current.status = status;
      if (FINAL_TOOL_STATUSES.has(status)) current.completedAt = now;
    }

    this.tools.set(toolCallId, current);
    this.trimRecentTools();
    return !wasSeen || statusChanged || (status !== undefined && FINAL_TOOL_STATUSES.has(status));
  }

  private trimRecentTools(): void {
    if (this.tools.size <= this.maxToolCalls) return;
    const ordered = [...this.tools.entries()].sort((a, b) => b[1].lastSeq - a[1].lastSeq);
    const keep = new Set(ordered.slice(0, this.maxToolCalls).map(([id]) => id));
    for (const id of this.tools.keys()) {
      if (!keep.has(id)) this.tools.delete(id);
    }
  }

  private publish(state: AgentAttemptTelemetryState, force: boolean): void {
    if (!this.onTelemetry) return;
    const now = this.now();
    if (!force && now - this.lastPublishedAt < this.flushIntervalMs) return;
    this.lastPublishedAt = now;
    this.onTelemetry(this.snapshot(state));
  }
}

interface MutableToolCall extends AgentToolCallTelemetry {
  lastSeq: number;
}

export function upsertAgentAttemptTelemetry(
  current: AgentTelemetry | undefined,
  attempt: AgentAttemptTelemetry
): AgentTelemetry {
  const previous = current?.attempts.filter((item) => item.attempt !== attempt.attempt) ?? [];
  return {
    currentAttempt: attempt.attempt,
    attempts: [...previous, attempt].sort((a, b) => a.attempt - b.attempt)
  };
}

export function latestAgentAttemptTelemetry(telemetry: AgentTelemetry | undefined): AgentAttemptTelemetry | undefined {
  if (!telemetry) return undefined;
  return telemetry.attempts.find((attempt) => attempt.attempt === telemetry.currentAttempt)
    ?? telemetry.attempts[telemetry.attempts.length - 1];
}

function buildPreview(text: string, edgeBytes: number, artifactRef?: string): AgentIoPreview {
  const originalBytes = Buffer.byteLength(text);
  const maxBytes = edgeBytes * 2;
  if (originalBytes <= maxBytes) {
    return { preview: text, truncated: false, originalBytes, headBytes: originalBytes, artifactRef };
  }
  const bytes = Buffer.from(text);
  const head = bytes.subarray(0, edgeBytes).toString("utf8");
  const tail = bytes.subarray(Math.max(edgeBytes, bytes.length - edgeBytes)).toString("utf8");
  return {
    preview: `${head}\n[acpus truncated: originalBytes=${originalBytes}]\n${tail}`,
    truncated: true,
    originalBytes,
    headBytes: edgeBytes,
    tailBytes: edgeBytes,
    artifactRef
  };
}

function parseObject(line: string): Record<string, unknown> | undefined {
  try {
    return readRecord(JSON.parse(line));
  } catch {
    return undefined;
  }
}

function extractUpdate(obj: Record<string, unknown>): Record<string, unknown> | undefined {
  return readRecord(readRecord(obj.params)?.update);
}

function isToolUpdate(update: Record<string, unknown>): boolean {
  const kind = update.sessionUpdate;
  return kind === "tool_call" || kind === "tool_call_update";
}

function readContentText(value: unknown): string | undefined {
  if (typeof value === "string") return value;
  const record = readRecord(value);
  if (!record) return undefined;
  if (typeof record.text === "string") return record.text;
  const nested = readRecord(record.content);
  return typeof nested?.text === "string" ? nested.text : undefined;
}

function readString(value: unknown): string | undefined {
  return typeof value === "string" && value.trim().length > 0 ? value : undefined;
}

function readNonNegativeNumber(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) && value >= 0 ? value : undefined;
}

function readRecord(value: unknown): Record<string, unknown> | undefined {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? value as Record<string, unknown>
    : undefined;
}

File: crates/acpus-runtime/src/agent_telemetry.rs
use crate::{
    AgentAttemptTelemetry, AgentAttemptTelemetryState, AgentContextUsage, AgentIoPreview,
    AgentTelemetry, AgentTokenUsage, AgentToolCallTelemetry, AgentToolsTelemetry,
};
use chrono::Utc;
use serde_json::Value;
use std::{collections::BTreeMap, path::Path};

const PREVIEW_EDGE_BYTES: usize = 8 * 1024;
const MAX_TOOL_CALLS: usize = 200;

#[derive(Default)]
struct MutableToolCall {
    call: Option<AgentToolCallTelemetry>,
    last_seq: usize,
}

pub fn agent_attempt_telemetry(
    attempt: u32,
    input_text: &str,
    stdout: &str,
    response_text: &str,
    state: AgentAttemptTelemetryState,
    cwd: &Path,
) -> AgentAttemptTelemetry {
    agent_attempt_telemetry_with_refs(AgentAttemptTelemetryInput {
        attempt,
        input_text,
        stdout,
        response_text,
        state,
        cwd,
        input_artifact_ref: None,
        output_artifact_ref: None,
        acpx_record_id: None,
    })
}

pub struct AgentAttemptTelemetryInput<'a> {
    pub attempt: u32,
    pub input_text: &'a str,
    pub stdout: &'a str,
    pub response_text: &'a str,
    pub state: AgentAttemptTelemetryState,
    pub cwd: &'a Path,
    pub input_artifact_ref: Option<String>,
    pub output_artifact_ref: Option<String>,
    pub acpx_record_id: Option<String>,
}

pub fn agent_attempt_telemetry_with_refs(
    input: AgentAttemptTelemetryInput<'_>,
) -> AgentAttemptTelemetry {
    let AgentAttemptTelemetryInput {
        attempt,
        input_text,
        stdout,
        response_text,
        state,
        cwd,
        input_artifact_ref,
        output_artifact_ref,
        acpx_record_id,
    } = input;
    let mut response = String::new();
    let mut context: Option<AgentContextUsage> = None;
    let mut token_usage = None;
    let mut tools: BTreeMap<String, MutableToolCall> = BTreeMap::new();
    let mut total_tool_call_count = 0usize;
    let mut seq = 0usize;
    let started_at = Utc::now().to_rfc3339();

    for line in stdout.lines() {
        let Ok(value) = serde_json::from_str::<Value>(line.trim()) else {
            continue;
        };
        if value.get("id").is_some()
            && value.get("result").is_some()
            && let Some(usage) = value.pointer("/result/usage").and_then(Value::as_object)
        {
            token_usage = read_token_usage(usage).or(token_usage);
        }
        let Some(update) = value.pointer("/params/update").and_then(Value::as_object) else {
            continue;
        };
        let updated_at = Utc::now().to_rfc3339();
        match update.get("sessionUpdate").and_then(Value::as_str) {
            Some("agent_message_chunk") => {
                if let Some(text) = read_content_text(update.get("content")) {
                    response.push_str(text);
                }
            }
            Some("usage_update") | Some("context_size") => {
                let used = update
                    .get("used")
                    .or_else(|| {
                        update
                            .get("content")
                            .and_then(|value| value.pointer("/used"))
                    })
                    .and_then(Value::as_u64);
                let size = update
                    .get("size")
                    .or_else(|| {
                        update
                            .get("content")
                            .and_then(|value| value.pointer("/size"))
                    })
                    .and_then(Value::as_u64);
                if let (Some(used), Some(size)) = (used, size) {
                    let used = if used == 0 {
                        context
                            .as_ref()
                            .filter(|context| context.used > 0)
                            .map(|context| context.used)
                            .unwrap_or(used)
                    } else {
                        used
                    };
                    context = Some(AgentContextUsage {
                        used,
                        size,
                        updated_at,
                    });
                }
            }
            Some("tool_call") | Some("tool_call_update") => {
                if let Some(tool_call_id) = non_empty_str(update.get("toolCallId")) {
                    let is_new = !tools.contains_key(tool_call_id);
                    if is_new {
                        total_tool_call_count += 1;
                    }
                    seq += 1;
                    let entry = tools.entry(tool_call_id.to_string()).or_default();
                    let mut call = entry.call.take().unwrap_or_else(|| AgentToolCallTelemetry {
                        tool_call_id: tool_call_id.to_string(),
                        title: None,
                        status: None,
                        kind: None,
                        tool_name: None,
                        started_at: updated_at.clone(),
                        updated_at: updated_at.clone(),
                        completed_at: None,
                    });
                    call.updated_at = updated_at.clone();
                    if let Some(title) = non_empty_str(update.get("title")) {
                        call.title = Some(title.to_string());
                    }
                    if let Some(kind) = non_empty_str(update.get("kind")) {
                        call.kind = Some(kind.to_string());
                    }
                    if let Some(status) = non_empty_str(update.get("status")) {
                        call.completed_at = final_tool_status(status).then_some(updated_at);
                        call.status = Some(status.to_string());
                    }
                    if let Some(tool_name) = non_empty_str(
                        update
                            .get("_meta")
                            .and_then(|value| value.pointer("/claudeCode/toolName")),
                    ) {
                        call.tool_name = Some(tool_name.to_string());
                    }
                    entry.call = Some(call);
                    entry.last_seq = seq;
                }
            }
            _ => {}
        }
    }

    let completed_at = matches!(
        state,
        AgentAttemptTelemetryState::Completed
            | AgentAttemptTelemetryState::Failed
            | AgentAttemptTelemetryState::Paused
            | AgentAttemptTelemetryState::Cancelled
    )
    .then(|| Utc::now().to_rfc3339());
    let mut recent_calls = tools
        .into_values()
        .filter_map(|entry| entry.call.map(|call| (entry.last_seq, call)))
        .collect::<Vec<_>>();
    recent_calls.sort_by_key(|entry| std::cmp::Reverse(entry.0));
    let dropped_tool_call_count = recent_calls.len().saturating_sub(MAX_TOOL_CALLS);
    recent_calls.truncate(MAX_TOOL_CALLS);

    AgentAttemptTelemetry {
        attempt,
        state,
        started_at,
        updated_at: completed_at
            .clone()
            .unwrap_or_else(|| Utc::now().to_rfc3339()),
        completed_at,
        context,
        token_usage,
        input: Some(build_preview(input_text, input_artifact_ref)),
        output: (!response_text.is_empty()).then(|| {
            build_preview(
                if response.is_empty() {
                    response_text
                } else {
                    response.as_str()
                },
                output_artifact_ref,
            )
        }),
        tools: AgentToolsTelemetry {
            total_tool_call_count,
            dropped_tool_call_count,
            recent_calls: recent_calls.into_iter().map(|(_, call)| call).collect(),
        },
        acpx_record_id,
        cwd: Some(cwd.display().to_string()),
    }
}

pub fn upsert_agent_attempt_telemetry(
    current: Option<AgentTelemetry>,
    mut attempt: AgentAttemptTelemetry,
) -> AgentTelemetry {
    let mut attempts = current
        .map(|telemetry| telemetry.attempts)
        .unwrap_or_default()
        .into_iter()
        .filter(|item| {
            if item.attempt == attempt.attempt {
                attempt.started_at.clone_from(&item.started_at);
                return false;
            }
            true
        })
        .collect::<Vec<_>>();
    attempts.push(attempt);
    attempts.sort_by_key(|attempt| attempt.attempt);
    AgentTelemetry {
        current_attempt: attempts.last().map(|attempt| attempt.attempt).unwrap_or(1),
        attempts,
    }
}

fn read_content_text(value: Option<&Value>) -> Option<&str> {
    let value = value?;
    value
        .as_str()
        .or_else(|| value.pointer("/text").and_then(Value::as_str))
        .or_else(|| value.pointer("/0/text").and_then(Value::as_str))
        .or_else(|| value.pointer("/content/text").and_then(Value::as_str))
}

fn read_token_usage(map: &serde_json::Map<String, Value>) -> Option<AgentTokenUsage> {
    let usage = AgentTokenUsage {
        source: "prompt_response".to_string(),
        input_tokens: read_u64(map, &["inputTokens", "input_tokens"]),
        output_tokens: read_u64(map, &["outputTokens", "output_tokens"]),
        cached_read_tokens: read_u64(
            map,
            &[
                "cachedReadTokens",
                "cacheReadInputTokens",
                "cache_read_input_tokens",
            ],
        ),
        cached_write_tokens: read_u64(
            map,
            &[
                "cachedWriteTokens",
                "cacheCreationInputTokens",
                "cache_creation_input_tokens",
            ],
        ),
        thought_tokens: read_u64(map, &["thoughtTokens", "thought_tokens"]),
        total_tokens: read_u64(map, &["totalTokens", "total_tokens"]),
    };
    (usage.input_tokens.is_some()
        || usage.output_tokens.is_some()
        || usage.cached_read_tokens.is_some()
        || usage.cached_write_tokens.is_some()
        || usage.thought_tokens.is_some()
        || usage.total_tokens.is_some())
    .then_some(usage)
}

fn read_u64(map: &serde_json::Map<String, Value>, keys: &[&str]) -> Option<u64> {
    keys.iter()
        .find_map(|key| map.get(*key).and_then(Value::as_u64))
}

fn non_empty_str(value: Option<&Value>) -> Option<&str> {
    value
        .and_then(Value::as_str)
        .filter(|value| !value.trim().is_empty())
}

fn final_tool_status(status: &str) -> bool {
    matches!(status, "completed" | "failed" | "cancelled")
}

fn build_preview(text: &str, artifact_ref: Option<String>) -> AgentIoPreview {
    let original_bytes = text.len();
    if original_bytes <= PREVIEW_EDGE_BYTES * 2 {
        return AgentIoPreview {
            preview: text.to_string(),
            truncated: false,
            original_bytes,
            head_bytes: original_bytes,
            tail_bytes: None,
            artifact_ref,
        };
    }
    let head_end = floor_char_boundary(text, PREVIEW_EDGE_BYTES);
    let tail_start = ceil_char_boundary(text, original_bytes - PREVIEW_EDGE_BYTES);
    AgentIoPreview {
        preview: format!(
            "{}\n[acpus truncated: originalBytes={}]\n{}",
            &text[..head_end],
            original_bytes,
            &text[tail_start..]
        ),
        truncated: true,
        original_bytes,
        head_bytes: PREVIEW_EDGE_BYTES,
        tail_bytes: Some(PREVIEW_EDGE_BYTES),
        artifact_ref,
    }
}

fn floor_char_boundary(text: &str, index: usize) -> usize {
    let mut index = index.min(text.len());
    while !text.is_char_boundary(index) {
        index -= 1;
    }
    index
}

fn ceil_char_boundary(text: &str, index: usize) -> usize {
    let mut index = index.min(text.len());
    while index < text.len() && !text.is_char_boundary(index) {
        index += 1;
    }
    index
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn captures_agent_chunks_tools_context_and_tokens() {
        let stdout = [
            r#"{"jsonrpc":"2.0","method":"session/update","params":{"update":{"sessionUpdate":"agent_message_chunk","content":{"type":"text","text":"hello "}}}}"#,
            r#"{"jsonrpc":"2.0","method":"session/update","params":{"update":{"sessionUpdate":"usage_update","used":25000,"size":190000}}}"#,
            r#"{"jsonrpc":"2.0","method":"session/update","params":{"update":{"sessionUpdate":"tool_call","toolCallId":"t1","title":"Read","status":"running","_meta":{"claudeCode":{"toolName":"Read"}}}}}"#,
            r#"{"jsonrpc":"2.0","method":"session/update","params":{"update":{"sessionUpdate":"tool_call_update","toolCallId":"t1","status":"completed"}}}"#,
            r#"{"jsonrpc":"2.0","method":"session/update","params":{"update":{"sessionUpdate":"agent_message_chunk","content":{"type":"text","text":"world"}}}}"#,
            r#"{"jsonrpc":"2.0","id":1,"result":{"usage":{"total_tokens":12000}}}"#,
        ]
        .join("\n");

        let telemetry = agent_attempt_telemetry(
            1,
            "prompt",
            &stdout,
            "hello world",
            AgentAttemptTelemetryState::Completed,
            Path::new("/tmp/work"),
        );

        assert_eq!(telemetry.output.as_ref().unwrap().preview, "hello world");
        assert_eq!(telemetry.context.as_ref().unwrap().used, 25000);
        assert_eq!(
            telemetry.token_usage.as_ref().unwrap().total_tokens,
            Some(12000)
        );
        assert_eq!(telemetry.tools.total_tool_call_count, 1);
        assert_eq!(
            telemetry.tools.recent_calls[0].title.as_deref(),
            Some("Read")
        );
        assert_eq!(
            serde_json::to_value(upsert_agent_attempt_telemetry(None, telemetry)).unwrap()["currentAttempt"],
            json!(1)
        );
    }

    #[test]
    fn ignores_result_like_payloads_without_json_rpc_id() {
        let telemetry = agent_attempt_telemetry(
            1,
            "prompt",
            r#"{"jsonrpc":"2.0","result":{"usage":{"inputTokens":10,"totalTokens":10}}}"#,
            "",
            AgentAttemptTelemetryState::Running,
            Path::new("/tmp/work"),
        );

        assert!(telemetry.token_usage.is_none());
    }

    #[test]
    fn upsert_preserves_started_at_for_same_attempt_updates() {
        let first = agent_attempt_telemetry(
            1,
            "prompt",
            "",
            "",
            AgentAttemptTelemetryState::Running,
            Path::new("/tmp/work"),
        );
        let started_at = first.started_at.clone();
        let mut second = agent_attempt_telemetry(
            1,
            "prompt",
            r#"{"jsonrpc":"2.0","method":"session/update","params":{"update":{"sessionUpdate":"agent_message_chunk","content":{"type":"text","text":"done"}}}}"#,
            "done",
            AgentAttemptTelemetryState::Completed,
            Path::new("/tmp/work"),
        );
        second.started_at = "later".to_string();

        let telemetry = upsert_agent_attempt_telemetry(
            Some(upsert_agent_attempt_telemetry(None, first)),
            second,
        );

        assert_eq!(telemetry.attempts[0].started_at, started_at);
        assert_eq!(
            telemetry.attempts[0].state,
            AgentAttemptTelemetryState::Completed
        );
    }

    #[test]
    fn preserves_nonzero_context_used_when_later_update_is_zero() {
        let stdout = [
            r#"{"jsonrpc":"2.0","method":"session/update","params":{"update":{"sessionUpdate":"usage_update","used":0,"size":200000}}}"#,
            r#"{"jsonrpc":"2.0","method":"session/update","params":{"update":{"sessionUpdate":"usage_update","used":40000,"size":200000}}}"#,
            r#"{"jsonrpc":"2.0","method":"session/update","params":{"update":{"sessionUpdate":"usage_update","used":0,"size":210000}}}"#,
        ]
        .join("\n");

        let telemetry = agent_attempt_telemetry(
            1,
            "prompt",
            &stdout,
            "",
            AgentAttemptTelemetryState::Running,
            Path::new("/tmp/work"),
        );

        let context = telemetry.context.unwrap();
        assert_eq!(context.used, 40000);
        assert_eq!(context.size, 210000);
    }

    #[test]
    fn allows_initial_zero_context_used() {
        let telemetry = agent_attempt_telemetry(
            1,
            "prompt",
            r#"{"jsonrpc":"2.0","method":"session/update","params":{"update":{"sessionUpdate":"usage_update","used":0,"size":200000}}}"#,
            "",
            AgentAttemptTelemetryState::Running,
            Path::new("/tmp/work"),
        );

        let context = telemetry.context.unwrap();
        assert_eq!(context.used, 0);
        assert_eq!(context.size, 200000);
    }

    #[test]
    fn includes_acpx_record_id_when_provided() {
        let telemetry = agent_attempt_telemetry_with_refs(AgentAttemptTelemetryInput {
            attempt: 1,
            input_text: "prompt",
            stdout: "",
            response_text: "",
            state: AgentAttemptTelemetryState::Running,
            cwd: Path::new("/tmp/work"),
            input_artifact_ref: None,
            output_artifact_ref: None,
            acpx_record_id: Some("mock-session-id".to_string()),
        });

        assert_eq!(telemetry.acpx_record_id.as_deref(), Some("mock-session-id"));
    }
}
